src/util/result.h

```diff
@@ -5,45 +5,65 @@
 
 #include <variant>
 
-/*
- * 'BResult' is a generic class useful for wrapping a return object
- * (in case of success) or propagating the error cause.
-*/
+namespace util {
+
+struct Error { bilingual_str message; };
+
+//! The util::Result class provides a standard way for functions to return
+//! either error messages or result values.
+//!
+//! It is intended for high-level functions that need to report error strings to
+//! end users. Lower-level functions that don't need this error-reporting and
+//! only need error-handling should avoid util::Result and instead use standard
+//! classes like std::optional, std::variant, and std::tuple, or custom structs
+//! and enum types to return function results.
+//!
+//! Usage examples can be found in \example ../test/result_tests.cpp, but in
+//! general code returning `util::Result<T>` values is very similar to code
+//! returning `std::optional<T>` values. Existing functions returning
+//! `std::optional<T>` can be updated to return `util::Result<T>` and return
+//! error strings usually just replacing `return std::nullopt;` with `return
+//! util::Error{error_string};`.
 template<class T>
-class BResult {
+class Result {
 private:
     std::variant<bilingual_str, T> m_variant;
 
+    template <typename FT>
+    friend class Result;
+    template <typename FT>
+    friend bilingual_str ErrorString(const Result<FT>& result);
+
 public:
-    BResult() : m_variant{Untranslated("")} {}
-    BResult(T obj) : m_variant{std::move(obj)} {}
-    BResult(bilingual_str error) : m_variant{std::move(error)} {}
-
-    /* Whether the function succeeded or not */
-    bool HasRes() const { return std::holds_alternative<T>(m_variant); }
-
-    /* In case of success, the result object */
-    const T& GetObj() const {
-        assert(HasRes());
-        return std::get<T>(m_variant);
-    }
-    T ReleaseObj()
-    {
-        assert(HasRes());
-        return std::move(std::get<T>(m_variant));
-    }
-
-    /* In case of failure, the error cause */
-    const bilingual_str& GetError() const {
-        assert(!HasRes());
-        return std::get<bilingual_str>(m_variant);
-    }
-
-    explicit operator bool() const { return HasRes(); }
+    Result(T obj) : m_variant{std::in_place_index_t<1>{}, std::move(obj)} {}
+    Result(Error error) : m_variant{std::in_place_index_t<0>{}, std::move(error.message)} {}
+
+    //! std::optional methods, so functions returning optional<T> can change to
+    //! return Result<T> with minimal changes to existing code, and vice versa.
+    bool has_value() const { return m_variant.index() == 1; }
+    const T& value() const LIFETIMEBOUND { assert(*this); return std::get<1>(m_variant); }
+    T& value() LIFETIMEBOUND { assert(*this); return std::get<1>(m_variant); }
+    explicit operator bool() const { return has_value(); }
+    const T* operator->() const { return &value(); }
+    const T& operator*() const LIFETIMEBOUND { return value(); }
+    T* operator->() { return &value(); }
```

Comment:
Was about to ask to add LIFETIMEBOUND, but for some reason it doesn't do anything for `->` :man_shrugging: 


```diff
diff --git a/src/test/result_tests.cpp b/src/test/result_tests.cpp
index 5737434b12..25de1eb6bc 100644
--- a/src/test/result_tests.cpp
+++ b/src/test/result_tests.cpp
@@ -89,7 +89,10 @@ BOOST_AUTO_TEST_CASE(check_value_or)
 {
     BOOST_CHECK_EQUAL(IntFn(10, true).value_or(20), 10);
     BOOST_CHECK_EQUAL(IntFn(10, false).value_or(20), 20);
-    BOOST_CHECK_EQUAL(NoCopyFn(10, true).value_or(20), 10);
+    const auto& a{NoCopyFn(10, true)->m_n};
+    BOOST_CHECK_EQUAL(*a, 10);
+    const auto& b{NoCopyFn(10, true).value().m_n};
+    BOOST_CHECK_EQUAL(*b, 10);
     BOOST_CHECK_EQUAL(NoCopyFn(10, false).value_or(20), 20);
 }
 
diff --git a/src/util/result.h b/src/util/result.h
index d038ffe133..cc1947e860 100644
--- a/src/util/result.h
+++ b/src/util/result.h
@@ -49,9 +47,9 @@ public:
     const T& value() const LIFETIMEBOUND { assert(*this); return std::get<1>(m_variant); }
     T& value() LIFETIMEBOUND { assert(*this); return std::get<1>(m_variant); }
     explicit operator bool() const { return has_value(); }
-    const T* operator->() const { return &value(); }
+    const T* operator->() const LIFETIMEBOUND { return &value(); }
     const T& operator*() const LIFETIMEBOUND { return value(); }
-    T* operator->() { return &value(); }
+    T* operator->() LIFETIMEBOUND { return &value(); }
     T& operator*() LIFETIMEBOUND { return value(); }
     template <class U>
     constexpr T value_or(U&& default_value) const& { return has_value() ? value() : std::forward<U>(default_value); }

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25721#discussion_r935541678

> Was about to ask to add LIFETIMEBOUND, but for some reason it doesn't do anything for `->` man_shrugging

Oh, I just omitted it because I didn't know LIFETIMEBOUND worked for pointers. Seems correct to add even if it doesn't produce any diagnostics for now though, so I added it.